rfcs/0000-support-injected-sidecars.md

```diff
@@ -0,0 +1,117 @@
+not work when you have init containers that need to reach out to the network. This is
+a [well documented][istio-init-container-compatability] technical limitation related to
+Istio's [automatic sidecar injection][istio-sidecar-injection]. The limitation is due to Istio adding a sidecar
+container to the kpack build pod which runs [envoy proxy][envoy-proxy] and sets up traffic from the build containers to
+be routed through the proxy. Due to the fact that init containers run before any normal container can start, kpack
+builds are unable to reach the network because they are attempting to communicate through the envoy proxy container that
+has not and will not start until the build completes.
+
+To get around this issue, kpack currently applies an annotation (`sidecar.istio.io/inject=false`) to all build pods to
+disable Istio sidecar injection. This presents a problem where users might want to take advantage of the various
+features of Istio, but cannot because kpack is bypassing it.
+
+[buildpack]: https://buildpacks.io/
+
+[cnb-lifecycle]: https://buildpacks.io/docs/concepts/components/lifecycle/
+
+[envoy-proxy]: https://www.envoyproxy.io/
+
+[init-containers]: https://kubernetes.io/docs/concepts/workloads/pods/init-containers/
+
+[istio]: https://istio.io/
+
+[istio-sidecar-injection]: https://istio.io/latest/docs/setup/additional-setup/sidecar-injection/#injection
+
+[istio-service-mesh]: https://istio.io/latest/about/service-mesh/#what-is-a-service-mesh
+
+[istio-init-container-compatability]: https://istio.io/latest/docs/setup/additional-setup/cni/#compatibility-with-application-init-containers
+
+## Outcome
+
+When kpack is running with `SUPPORT_INJECTED_SIDECARS=true`, builds can execute successfully in environments where
+sidecars are executed into the build pods (i.e. istio proxy sidecar injection).
+
+## Actions to take
+
+1. Create a new binary that can be used as a new entrypoint for build steps. The purpose of this binary will be to allow
+   for specific build steps to depend on other build steps to complete before executing since the containers will be
+   running concurrently. There will also need to be a mechanism for propagating errors in one build step so that a build
+   can exit on error.
+
+   Example of possible implementation is in the [spike][kpack-spike-build-waiter].
+2. Convert the existing init containers in to be regular containers and add modify the entrypoint to be the new binary.
+3. We will put this new behavior behind a feature gate (Environment variable added to kpack controller deployment
+   called `SUPPORT_INJECTED_SIDECARS`) and only execute in regular containers if enabled by the user. In
+   the event that the feature flag is not set or set to false, kpack will continue to execute builds in init containers.
+4. Due to the fact that sidecars such as istio-proxy, do not exit when all other containers exit, we will need to add a
+   method for stopping the proxy once a build exits
```

Comment:
Do we also want a mechanism to determine if side cars are ready? It it possible for a lifecycle step to execute before the envoy-proxy fully spins up?

Reply:
I actually did that in the proposed implementation, but I wasn't sure if it was too much of an implementation detail to mention in the rfc. If you think it's worth calling out then I can add it

Reply:
Personally, I feel like if there's a call out for the side car stopping mechanism, there should be a call out for the lifecycle starting mechanism